overlord/state/taskrunner.go

```diff
@@ -42,91 +40,137 @@ type TaskRunner struct {
+// run must be called with the state lock in place
+func (r *TaskRunner) run(t *Task) {
+	var handler HandlerFunc
+	switch t.Status() {
+	case DoStatus:
+		t.SetStatus(DoingStatus)
+		fallthrough
+	case DoingStatus:
+		handler = r.handlers[t.Kind()].do
+
+	case UndoStatus:
+		t.SetStatus(UndoingStatus)
+		fallthrough
+	case UndoingStatus:
+		handler = r.handlers[t.Kind()].undo
+
+	default:
+		panic("internal error: attempted to run task in status " + t.Status().String())
+	}
+	if handler == nil {
+		panic("internal error: attempted to run task with nil handler for status " + t.Status().String())
 	}
-}
 
-// run must be called with the state lock in place
-func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 	tomb := &tomb.Tomb{}
-	id := task.ID()
-	r.tombs[id] = tomb
+	r.tombs[t.ID()] = tomb
 	tomb.Go(func() error {
-		defer func() {
-			r.mu.Lock()
-			defer r.mu.Unlock()
-
-			delete(r.tombs, id)
-		}()
-
-		// capture the error result with tomb.Kill so we can
+		// Capture the error result with tomb.Kill so we can
 		// use tomb.Err uniformily to consider both it or a
 		// overriding previous Kill reason.
-		tomb.Kill(fn(task, tomb))
+		tomb.Kill(handler(t, tomb))
 
+		// Locks must be acquired in the same order everywhere.
+		r.mu.Lock()
+		defer r.mu.Unlock()
 		r.state.Lock()
 		defer r.state.Unlock()
+
+		delete(r.tombs, t.ID())
+
 		switch err := tomb.Err(); err {
 		case Retry:
-			// Do nothing. Handler asked to try again later.
-			// TODO: define how to control retry intervals,
-			// right now things will be retried at the next Ensure
+			// Handler asked to be called again later.
+			// TODO Allow postponing retries past the next Ensure.
+			if t.Status() == AbortStatus {
+				// Would work without it but might take two ensures.
+				r.swapAbort(t)
+			}
 		case nil:
-			task.SetStatus(DoneStatus)
-			if len(task.HaltTasks()) > 0 {
-				// give a chance to taskrunners Ensure to start
-				// the waiting ones
+			var next []*Task
+			switch t.Status() {
+			case DoingStatus:
+				t.SetStatus(DoneStatus)
+				next = t.HaltTasks()
+			case AbortStatus:
+				t.SetStatus(DoneStatus) // Not actually aborted.
+				r.swapAbort(t)
+			case UndoingStatus:
+				t.SetStatus(UndoneStatus)
+				next = t.WaitTasks()
+			}
+			if len(next) > 0 {
 				r.state.EnsureBefore(0)
 			}
 		default:
-			task.SetStatus(ErrorStatus)
-			task.Errorf("%s", err)
-			propagateError(task)
+			t.SetStatus(ErrorStatus)
+			t.Errorf("%s", err)
+			r.abortChange(t.Change())
 		}
+
 		return nil
 	})
 }
 
-// mustWait returns whether task t must wait for other tasks to be done.
-func mustWait(t *Task) bool {
-	for _, wt := range t.WaitTasks() {
-		if wt.Status() != DoneStatus {
-			return true
+func (r *TaskRunner) abortChange(chg *Change) {
+	chg.Abort()
+	scheduled := false
+	for _, t := range chg.Tasks() {
+		if t.Status() == AbortStatus {
+			if tb, ok := r.tombs[t.ID()]; ok {
+				tb.Kill(nil)
+			}
+		}
+		if !scheduled && !t.Status().Ready() {
```

Comment:
I idly wonder if it would make sense (and potentially easier to read) to make this two loops over chg.Tasks(), so that `scheduled` is not needed and we simply break in the second loop once we found the first !t.Status().Ready().